codegen/src/api/metadata.rs

```diff
@@ -0,0 +1,229 @@
+    set: &mut HashSet<u32>,
+) -> [u8; 32] {
+    let mut bytes = vec![MetadataHashableIDs::Field as u8];
+
+    bytes.extend(field.name().encode());
+    bytes.extend(field.type_name().encode());
+    bytes.extend(get_type_hash(registry, field.ty().id(), set));
+
+    hash(&bytes)
+}
+
+fn get_variant_hash(
+    registry: &PortableRegistry,
+    var: &Variant<PortableForm>,
+    set: &mut HashSet<u32>,
+) -> [u8; 32] {
+    let mut bytes = vec![MetadataHashableIDs::Variant as u8];
+
+    bytes.extend(var.name().as_bytes());
+    for field in var.fields() {
+        bytes.extend(get_field_hash(registry, field, set));
+    }
+
+    hash(&bytes)
+}
+
+fn get_type_def_hash(
+    registry: &PortableRegistry,
+    ty_def: &TypeDef<PortableForm>,
+    set: &mut HashSet<u32>,
+) -> [u8; 32] {
+    let mut bytes = vec![MetadataHashableIDs::TypeDef as u8];
+
+    let data = match ty_def {
+        TypeDef::Composite(composite) => {
+            let mut bytes = Vec::new();
+            for field in composite.fields() {
+                bytes.extend(get_field_hash(registry, field, set));
+            }
+            bytes
+        }
+        TypeDef::Variant(variant) => {
+            let mut bytes = Vec::new();
+            for var in variant.variants() {
+                bytes.extend(get_variant_hash(registry, var, set));
+            }
+            bytes
+        }
+        TypeDef::Sequence(sequence) => {
+            let mut bytes = Vec::new();
+            bytes.extend(get_type_hash(registry, sequence.type_param().id(), set));
+            bytes
+        }
+        TypeDef::Array(array) => {
+            let mut bytes = Vec::new();
+            bytes.extend(array.len().to_be_bytes());
+            bytes.extend(get_type_hash(registry, array.type_param().id(), set));
+            bytes
+        }
+        TypeDef::Tuple(tuple) => {
+            let mut bytes = Vec::new();
+            for field in tuple.fields() {
+                bytes.extend(get_type_hash(registry, field.id(), set));
+            }
+            bytes
+        }
+        TypeDef::Primitive(primitive) => {
+            let mut bytes = Vec::new();
+            bytes.extend(primitive.encode());
+            bytes
+        }
+        TypeDef::Compact(compact) => {
+            let mut bytes = Vec::new();
+            bytes.extend(get_type_hash(registry, compact.type_param().id(), set));
+            bytes
+        }
+        TypeDef::BitSequence(bitseq) => {
+            let mut bytes = Vec::new();
+            bytes.extend(get_type_hash(registry, bitseq.bit_order_type().id(), set));
+            bytes.extend(get_type_hash(registry, bitseq.bit_store_type().id(), set));
+            bytes
+        }
+    };
+    bytes.extend(data);
+    hash(&bytes)
+}
+
+fn get_type_hash(
+    registry: &PortableRegistry,
+    id: u32,
+    set: &mut HashSet<u32>,
+) -> [u8; 32] {
+    lazy_static! {
+        static ref CACHED_UID: Mutex<HashMap<u32, [u8; 32]>> = Mutex::new(HashMap::new());
+    }
```

Comment:
The issue with a `lazy_static` cache I think is that, since it persists for the lifetime of the program, it could end up being shared between different calls to `get_type_hash` with different PortableRegistries and, so, incompatible type IDs (I think this is a problem regardless of struct vs function approach).

The more I read this code though, the more I can see that the cache makes a lot of sense at this level, since I imagine there will be a lot of duplicated calls to the same type IDs. 

Possible approaches from here could be:
1. Allow a cache to be provided to the functions. It's up to the caller to use the correct cache then, and not reuse the same cache across different PortableRegistries. No lifetimes to worry about here; the cache can be stored away wherever makes sense to the caller. Probably the easiest approach. The cache could be "opaque" outside of this module as well, so that external parties can simply create one and pass it in, but can't inspect the inside of it).
2. Meet in the middle somewhere between the "struct that owns metadata" that you had before, and this fully functional approach now, and have something like `pub struct MetadataHasher<'a>{ metadata: &'a RuntimeMetadataLastVersion, cache: HashMap<u32, [u8; 32]> }` which keeps a cache as long as it's borrowed that specific metadata to generate hashes from. Might be a pain for lifetime reasons though!
3. Revert to your "struct that takes ownership of metadata" approach. It feels worse to me in the sense that it shouldn't need to own metadata, but it would def be easier to work with than (2).

What do you think is the best approach?

Reply:
The first option seems more appropriate in terms of simplicity and maintainability for now.

I'm also not a fan of the third option, as the metadata hasher might live for a limited amount of time and the underlying implementation requires only a ref. 😄 

While the second option seems more natural, it comes with the downside of having a `MetadataHasher<'a>` on the `subxt::Client`, which already has an instance of metadata as owned. At that point, we'll need to refactor the code to avoid self-referencing structs 😄